Components: Optimize withFilters, avoiding per-instance bindings

aduth opened the pull request:
Related: #12824

This pull request seeks to optimize the implementation of `withFilters`. In my testing of a large post consisting of 20,000 words (70 headings, 501 paragraphs), ~400ms real-time is spent during initialization in `addHook`, `runHooks`, and `FilteredComponent#constructor`. These changes bring this down to roughly 170ms, with `runHooks` being the primary contributor (160ms), as it's also used extensively elsewhere in blocks parsing (i.e. `withFilters` contribution approaches negligible).

Before|After
---|---
![Before](https://user-images.githubusercontent.com/1779930/50794953-b015e800-129a-11e9-8fd4-16650e184ffe.png)|![After](https://user-images.githubusercontent.com/1779930/50794958-b6a45f80-129a-11e9-9612-bad4abd58804.png)

**Implementation notes:**

The specific optimizations include:

- Binding at most one hook handler for `addFilter` and `removeFilter` per hook name. Previously, a handler was added for each instance of a filtered component.
- Creating a single shared filtered component definition per hook name. Previously, each instance would create their own filtered component.

There has been one implementation change: Previously, the hooks were bound in the component's constructor. It has been changed here to occur during `componentDidMount`. This is in accordance with React recommendations, and avoids potential issues with lingering subscriptions in a server-rendered environment (see also #11819).

>Avoid introducing any side-effects or subscriptions in the constructor. For those use cases, use `componentDidMount()` instead.

https://reactjs.org/docs/react-component.html#constructor

**Testing instructions:**

Repeat testing instructions from #4428, substituting the hook name `blocks.BlockEdit` with the since-renamed `editor.BlockEdit`. In other words, insert this code in your console:

```
wp.hooks.addFilter( 'editor.BlockEdit', 'my/filters', () => () => wp.element.createElement( 'h1', {}, 'My block' ) );
```

aduth committed:
Components: Use single hook delegator for withFilters

aduth committed:
Component: Share component definition for withFilters instances

aduth commented on packages/components/src/higher-order/with-filters/index.js:
```diff
@@ -24,45 +24,82 @@ const ANIMATION_FRAME_PERIOD = 16;
-					this.Component = applyFilters( hookName, OriginalComponent );
-					this.forceUpdate();
-				}, ANIMATION_FRAME_PERIOD );
+				this.throttledForceUpdate = debounce(
```
We could probably go one step further and move the debounce to a single shared function, calling each instance's `forceUpdate` directly. This has a couple advantages:

- The constructor can be dropped altogether
- Avoids need to cancel in `componentWillUnmount`
- Automatically accounts for unmounting of components, since the callback would only evaluate `FilteredComponentRenderer.instances` after the debounce delay

Since it requires a bit more reorganization, I'm a bit inclined to leave this for a subsequent pull request.

aduth committed:
Components: Use slash for withFilters subgrouping

gziolo reviewed (changes requested):
There are some failing e2e tests which look related to this change:
- annotations
- block alignment

I think they use `withFilters` behind the scenes. It should be further investigated.

gziolo commented on packages/components/src/higher-order/with-filters/index.js:
```diff
@@ -24,45 +24,82 @@ const ANIMATION_FRAME_PERIOD = 16;
-				addAction( 'hookAdded', this.namespace, this.onHooksUpdated );
+				// If there were previously no mounted instances for components
+				// filtered on this hook, add the hook handler.
+				if ( FilteredComponentRenderer.instances.length === 1 ) {
```
Thinking a bit more about it, it is probably correct.

aduth replied:
> Thinking a bit more about it, it is probably correct. I mean, that when mounting it always increments the counter.

Yes, being that it only increments here, I think it should be fine as it is.

gziolo commented on packages/components/src/higher-order/with-filters/index.js:
```diff
@@ -24,45 +24,82 @@ const ANIMATION_FRAME_PERIOD = 16;
+		 *
+		 * @type {Component}
+		 */
+		let FilteredComponent = applyFilters( hookName, OriginalComponent );
```
I suspect this line might be the reason for failing e2e tests.

Before, we would call it in the constructor, which is much later in the flow - as it is after the editor starts to render. In this case, it happens just after HOCs are applied - which happens just after the component is exported from the JS module.

aduth replied:
> I suspect this line might be the reason for failing e2e tests.
> 
> Before, we would call it in the constructor, which is much later in the flow - as it is after the editor starts to render. In this case, it happens just after HOCs are applied - which happens just after the component is exported from the JS module.

That makes sense. Maybe then we can have this be returned from some getter function which computes the value once when it's called.

Like Lodash's `once`:

https://lodash.com/docs/4.17.11#once

aduth commented on packages/components/src/higher-order/with-filters/index.js:
```diff
@@ -24,45 +24,82 @@ const ANIMATION_FRAME_PERIOD = 16;
-				this.Component = applyFilters( hookName, OriginalComponent );
-				this.namespace = uniqueId( 'core/with-filters/component-' );
-				this.throttledForceUpdate = debounce( () => {
-					this.Component = applyFilters( hookName, OriginalComponent );
```
> @aduth, I saw this question in my inbox. I can't see it anymore in the original PR. I just wanted to confirm that it is triggered here and in your optimized version properly.

Yes, I deleted it once I noticed I'd overlooked the re-assignment in the debounce callback.

> The benefit of calling `applyFilters` inside of the `throttledForceUpdate` method is that it will be executed only once when multiple calls of add/remove filters occurs in the same small period of time.

Yeah, I can see that being a good thing. It might require that we do the refactoring I'd pondered at https://github.com/WordPress/gutenberg/pull/13231#discussion_r245809765 ?

aduth commented on packages/components/src/higher-order/with-filters/index.js:
```diff
@@ -24,45 +24,82 @@ const ANIMATION_FRAME_PERIOD = 16;
+			}
+
+			// Recreate the filtered component.
+			FilteredComponent = applyFilters( hookName, OriginalComponent );
```
Moving this back to  `throttledForceUpdate` would mean each component re-evaluates the `applyFilters` itself, rather than doing it just once for the shared definition. This is what I meant by "requiring the refactor" in https://github.com/WordPress/gutenberg/pull/13231#discussion_r246045044, since if we handled the debounce outside the components, it would only call `applyFilters` once, after the delay.

aduth committed:
Components: Assign FilteredComponent only once renderer constructed

aduth committed:
Component: Handle withFilters forced update through delegator

Ensures that FilteredComponent definition is only computed once per animation frame debounce

aduth commented:
The tests appear to be passing now with the latest revisions, which include (a) not computing the filtered component definition until the first instance of the filtered component is constructed and (b) handling the throttling outside the component class so that the application of filters applies only once, and only after the debounce delay.

gziolo commented:
I tested with the following snippet:

```js
wp.hooks.addFilter( 'blocks.BlockEdit', 'my/filters/1', () => () => wp.element.createElement( 'h1', {}, 'My block 1' ) );
```

and 

```js
wp.hooks.removeFilter( 'editor.BlockEdit', 'my/filters/1' )
```

gziolo reviewed (approved):
This looks great and tests well. Awesome work 🚢

adamsilverstein commented:
Excellent!
